mlir_graphblas/engine.py

```diff
@@ -545,6 +547,80 @@ def __init__(self, cli_executable=None, cli_options=None, llvmlite_engine=None):
         self.name_to_callable: Dict[str, Callable] = {}
         return
 
+    def profiled_function(self, main_callable: Callable) -> Callable:
+        """Decorator to profile a function via Linux's perf tool."""
+
+        def mp_func(queue: mp.SimpleQueue, *args):
+            # send pid to decorated_func
+            queue.put(os.getpid())
+            # wait for decorated_func to signal start of execution
+            start_signal = queue.get()
+            assert start_signal == "start"
+            # execute the function
+            result = main_callable(*args)
+            # return the result on the queue
+            queue.put(result)
+            return
+
+        def decorated_func(*args) -> Any:
+            # get pid from mp_func
+            q = mp.SimpleQueue()
+            execution_process = mp.Process(target=mp_func, args=(q, *args))
+            execution_process.start()
+            execution_process_id = q.get()
+
+            # start profiling
+            record_process = subprocess.Popen(
+                "/bin/bash",
+                stdin=subprocess.PIPE,
+                stdout=subprocess.DEVNULL,
+                stderr=subprocess.PIPE,
+            )
+            record_process.stdin.write(
+                f"""
+            perf record -p {execution_process_id}
+            exit
+            """.encode()
+            )
+            record_process.stdin.flush()
+
+            # wait for profiling to initialize
+            for _ in range(3):  # FIXME how do we truly know perf-record is ready?
+                record_process.stderr.readline()
```

Comment:
I don't know of a better way to know when [perf-record](https://man7.org/linux/man-pages/man1/perf-record.1.html) is in a state where we can safely start executing the function.

Reply:
Stuart mentioned that he had to sleep for a second or two, since there isn't a good way to tell if it is ready.

Reply:
I'll leave this as-is for now. If things get buggy, it might be a better-but-not-great idea to use a sleep instead.